### PR TITLE
fix/16098: Listen keydown from user to focus again the composer

### PR DESCRIPTION
**Pull request by tienifr**
### Details

We should automatically focus on the Composer if we press on regular characters like a, b, c, d, ... (not non-character keys like Enter, Shift) when we're not typing on any input or textarea.

### Fixed Issues

$ https://github.com/Expensify/App/issues/16098
PROPOSAL: https://github.com/Expensify/App/issues/16098#issuecomment-1485325162

### Tests

1. Go to any chats
2. Click on any message
3. Press any character keys like a,b,c,d (not non-character keys like Enter, Shift) 
4. Verify that the composer is focused again and the above character is entered in the composer

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Go to any chats
2. Click on any message
3. Press any character keys like a,b,c,d (not non-character keys like Enter, Shift) 
4. Verify that the composer is focused again and the above character is entered in the composer
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/113963320/230654500-219234f4-f58f-4a03-9bff-036b345a40db.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/113963320/230656324-fb443574-697d-47b1-99ea-d7ac319f2ea5.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/113963320/230656303-1a0cd2b2-7ae6-46af-8869-41c4f3be5f66.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/113963320/230656644-7dd4d41e-5123-4cc4-bcff-ae1e61f503f1.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/113963320/230657543-79dbff62-e0c3-4bb1-933c-c296b99f7006.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/113963320/230659268-9e2b2969-22e2-4454-9b66-79631be61369.mov

</details>

### CODE REVIEW

**Comment by PauloGasparSv** [comment]
Won't be able to review this Today but will prioritize it on Monday!

**Comment by rushatgabhane** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

File: src/pages/home/report/ReportActionCompose/listenKeyDown.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+function listenKeyDown(callback) {
+    document.addEventListener('keydown', callback);
```
Comment (rushatgabhane) [inline src/pages/home/report/ReportActionCompose/listenKeyDown.js:2]: @tienifr nitpick: style change -

condense `listenKeyDown` and `removeListenKeyDown` to a single file.

File: src/pages/home/report/ReportActionCompose/index.js [hunk 2]
```diff
@@ -460,23 +467,42 @@ class ReportActionCompose extends React.Component {
...
     }
 
+    keydownListener(e) {
+        if (this.state.isFocused) { return; }
+
+        // if the key pressed is non-character keys like Enter, Shift, ... do not focus
+        if (e.key.length > 1) { return; }
+
+        // if we're typing on another input/text area, do not focus
+        if (e.target.nodeName === 'INPUT' || e.target.nodeName === 'TEXTAREA') { return; }
```
Comment (rushatgabhane) [inline src/pages/home/report/ReportActionCompose/index.js:477]: nitpick: I don't have anything against retuning on the same line, but it isn't consistent with the codebase.

```suggestion
        if (e.target.nodeName === 'INPUT' || e.target.nodeName === 'TEXTAREA') { 
            return; 
        }
```

File: src/pages/home/report/ReportActionCompose/index.js [hunk 3]
```diff
@@ -255,6 +261,7 @@ class ReportActionCompose extends React.Component {
     }
 
     componentWillUnmount() {
+        removeListenKeyDown(this.keydownListener);
```
Comment (rushatgabhane) [inline src/pages/home/report/ReportActionCompose/index.js:264]: We should perform this only if `keydownListener` is not null
```suggestion
        if (this.keydownListener) {
            removeListenKeyDown(this.keydownListener);
```

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
@tienifr nice polish to the UX!

Left a few minor style change requests

**Review by rushatgabhane — DISMISSED** [review]
@PauloGasparSv LGTM! 🎉 

appreciate the quick work @tienifr 🚀

**Comment by PauloGasparSv** [comment]
Hey, the changes here are looking great!!! I'm testing on all platforms here too.

I'm testing if it's ok to introduce the following behavior though. In **Web/Desktop** if we have a chat open and we access the Profile Settings (or any other "page") and start typing stuff it will focus on the composer and add the input there.

That makes total sense but may be confusing to the user, for example in Desktop Web:

https://user-images.githubusercontent.com/6564265/230911203-3814025d-d700-4c69-9b1a-e97bcdd49234.mov

And Desktop Web with a small screen (mobile mode). This one may be a problem but I don't think it's worth solving since no one will probably use the Web version like this. 

https://user-images.githubusercontent.com/6564265/230911439-50804625-dc74-4ae3-a7d7-56875010cd38.mov

I'm checking if this happens on mWeb too and if there is a simple fix for this. If there isn't IMO we should send a message in #expensify-open-source to get other people's opinions (mine is Do Nothing).

**Comment by tienifr** [comment]
@rushatgabhane Please help to check my small fix for the above problem

**Comment by rushatgabhane** [comment]
@tienifr instead of checking visibility inside component did update can we modify the callback function for the event listener to return early if modal isn't visible?

**Comment by tienifr** [comment]
@rushatgabhane Your suggestion is so great. Just updated

**Comment by PauloGasparSv** [comment]
Amazing!!! Re-testing here in a bit!

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
@tienifr non-blocking bug: 
1. Go to mWeb
2. Open LHN
3. Type "123"
4. Go to the most recent chat

Typing while on LHN page in mWeb types into the composer

@tienifr is there an easy way to find if LHN is being focused?

https://user-images.githubusercontent.com/29673073/231001610-27d7978d-87f3-4ff9-9e9c-7d68eae210b6.mov

**Review by rushatgabhane — DISMISSED** [review]
@PauloGasparSv nice catch with the bug related to RHN 

LGTM! 

https://user-images.githubusercontent.com/29673073/231001169-e981db04-ac11-401c-911f-f5a19195de21.mov


https://user-images.githubusercontent.com/29673073/231001190-00bb4d59-a9b5-48ab-b4d7-9658887523cd.mov

**Comment by PauloGasparSv** [comment]
I'll have to re-review this tomorrow as I'm still focused on another important issue!
Will prioritize it and do it in the morning : )

**Comment by tienifr** [comment]
@rushatgabhane Just updated, we should disable keydownListen in the small screen (mWeb and native) on the callback function

**Comment by rushatgabhane** [comment]
That's an acceptable trade-off right? @PauloGasparSv 
This feature is disabled for mWeb

**Review by rushatgabhane — DISMISSED** [review]

**Comment by PauloGasparSv** [comment]
Hey @tienifr I took to long to review and now we have conflicts with main, sorry for that!!!!

I just finished reviewing all platforms and was about to approve this but I found 1 last bug! If we have the emoji search opened and remove the focus from it's search then: both the emoji search input and the message composer listen for keyboard input:

https://user-images.githubusercontent.com/6564265/231210203-267cbc4b-8222-4870-a7a9-9a1c04e95960.mov


Do you mind taking a look at it?

**Comment by tienifr** [comment]
@PauloGasparSv I added a new state `isEmojiPickerVisible` to check if the emoji Picker is open and if `isEmojiPickerVisible` is true, we will not trigger the `keydownListener` callback function
cc @rushatgabhane

**Review by rushatgabhane — APPROVED** [review]

**Comment by tienifr** [comment]
@PauloGasparSv Just updated

**Review by PauloGasparSv — APPROVED** [review]
LGTM!
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/6564265/231529042-ff958839-fea4-493a-9533-28bc222b8da3.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/6564265/231530702-d381f0a9-3656-45fc-9571-7b94fbb054f3.mov

</details>

<details>
<summary>iOS Native</summary>

https://user-images.githubusercontent.com/6564265/231538034-2b57cec8-b3b2-4153-9d39-409f37272e26.mov



</details>
<details>
<summary>iOS mWeb</summary>


https://user-images.githubusercontent.com/6564265/231538639-bbd19173-1e44-465a-98d5-914031764f12.mov


</details>

<details>
<summary>Android Native</summary>

https://user-images.githubusercontent.com/6564265/231543644-a907ef2b-b294-45fb-9679-95b4c45b64a2.mov

</details>
<details>
<summary>Android mWeb</summary>

https://user-images.githubusercontent.com/6564265/231541277-d9dd6604-fe2a-4b27-a99c-438392af3f67.mov

</details>

**Comment by PauloGasparSv** [comment]
Hey @tienifr can you also mention in the QA steps that the behaviour should only work for Desktop and Desktop Web?

**Comment by PauloGasparSv** [comment]
Taking a look at that, I think everyone else is out right now.

**Comment by PauloGasparSv** [comment]
>  There are many regressions being reported from this PR. Can we please revert this @rushatgabhane?

Was about to comment that after seeing [this](https://github.com/Expensify/App/pull/17428#issuecomment-1509014993).

**Comment by PauloGasparSv** [comment]
We should also revert https://github.com/Expensify/App/pull/17360

**Comment by PauloGasparSv** [comment]
Bump @tienifr too, can you take a look at this and help revert both P.R.'s? 
- https://github.com/Expensify/App/pull/17138
- https://github.com/Expensify/App/pull/17360

**Pull request merged**